internal/notifier/notifier.go

```diff
@@ -24,189 +28,327 @@ const (
-	notificationBuf  chan *riverdriver.Notification
-	statusChangeFunc func(componentstatus.Status)
-	logger           *slog.Logger
-
-	mu           sync.Mutex
-	isConnActive bool
-	subs         map[NotificationTopic][]*Subscription
-	subChangeCh  chan *subscriptionChange
+	startstop.BaseStartStop
+
+	listener          riverdriver.Listener
+	notificationBuf   chan *riverdriver.Notification
+	statusChangeFunc  func(componentstatus.Status)
+	testSignals       notifierTestSignals
+	waitInterruptChan chan func()
+
+	mu            sync.RWMutex
+	isConnected   bool
+	isStarted     bool
+	isWaiting     bool
+	subscriptions map[NotificationTopic][]*Subscription
+	waitCancel    context.CancelFunc
 }
 
-func New(archetype *baseservice.Archetype, listener riverdriver.Listener, statusChangeFunc func(componentstatus.Status), logger *slog.Logger) *Notifier {
+func New(archetype *baseservice.Archetype, listener riverdriver.Listener, statusChangeFunc func(componentstatus.Status)) *Notifier {
 	notifier := baseservice.Init(archetype, &Notifier{
-		listener:         listener,
-		notificationBuf:  make(chan *riverdriver.Notification, 1000),
-		statusChangeFunc: statusChangeFunc,
-		logger:           logger.WithGroup("notifier"),
+		listener:          listener,
+		notificationBuf:   make(chan *riverdriver.Notification, 1000),
+		statusChangeFunc:  statusChangeFunc,
+		waitInterruptChan: make(chan func(), 10),
 
-		subs:        make(map[NotificationTopic][]*Subscription),
-		subChangeCh: make(chan *subscriptionChange, 1000),
+		subscriptions: make(map[NotificationTopic][]*Subscription),
 	})
 	return notifier
 }
 
-func (n *Notifier) Run(ctx context.Context) {
-	n.statusChangeFunc(componentstatus.Initializing)
-	var wg sync.WaitGroup
+func (n *Notifier) Start(ctx context.Context) error {
+	ctx, shouldStart, stopped := n.StartInit(ctx)
+	if !shouldStart {
+		return nil
+	}
+
+	// The loop below will connect/close on every iteration, but do one initial
+	// connect so the notifier fails fast in case of an obvious problem.
+	if err := n.listenerConnect(ctx, false); err != nil {
+		close(stopped)
+		if errors.Is(err, context.Canceled) {
+			return nil
+		}
+		return err
+	}
 
-	wg.Add(1)
 	go func() {
-		defer wg.Done()
-		n.deliverNotifications(ctx)
-	}()
+		defer close(stopped)
 
-	for {
-		n.getConnAndRun(ctx)
+		n.Logger.InfoContext(ctx, n.Name+": Notifier started")
+		defer n.Logger.InfoContext(ctx, n.Name+": Notifier stopped")
 
-		select {
-		case <-ctx.Done():
-			wg.Wait()
-			n.logger.Info(n.Name + ": Notifier stopped")
-			n.statusChangeFunc(componentstatus.Stopped)
-			return
-		default:
-			// TODO: exponential backoff
+		n.withLock(func() { n.isStarted = true })
+		defer n.withLock(func() { n.isStarted = false })
+
+		defer n.listenerClose(ctx, false)
+
+		n.statusChangeFunc(componentstatus.Initializing)
+		var wg sync.WaitGroup
+
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			n.deliverNotifications(ctx)
+		}()
+
+		for attempt := 0; ; attempt++ {
```

Comment:
This is taken care of by the base service:

``` go
// Reset exponential backoff counters after a certain point so that sleep
// durations aren't flung into a ridiculously distant future.
const exponentialBackoffAttemptResetAfter = 10

func (s *BaseService) exponentialBackoffSecondsWithoutJitter(attempt int) float64 {
	// We use a different exponential backoff algorithm here compared to the
	// default retry policy (2**N versus N**4) because it results in more
	// retries sooner. When it comes to exponential backoffs in services we
	// never want to sleep for hours/days, unlike with failed jobs.
	return math.Pow(2, float64(attempt%exponentialBackoffAttemptResetAfter))
}
```

I went back and forth on where best to implement it, but putting it in the base has the major advantages of (1) it being in one place and impossible to forget, and (2) more easily testable to make sure the right thing is happening.

Reply:
Ah, the max before reset idea is that it it'll backoff totally normally, but after hitting 10 attempts, the sleep schedule resets. So there's no quick burst — it sleeps on a normal schedule, but makes sure to reset before the sleep durations get too obscene.

Just to make this more obvious, I've retooled the API to now look like this:

``` go
s.CancellableSleepExponentialBackoff(ctx, attempt, baseservice.MaxAttemptsBeforeResetDefault)
```

So the reset is still built in, but it's now visible in the function invocation that a reset is a thing, which will hopefully improve legibility. Not necessarily the be all end all if this turns out to be suboptimal — will keep an eye on it to make sure it continues to be reasonable.